clinto/parsers/base.py

```diff
@@ -0,0 +1,63 @@
+from __future__ import absolute_import
+import json
+import os
+import copy
+
+
+def update_dict_copy(a, b):
+    temp = copy.deepcopy(a)
+    temp.update(b)
+    return temp
+
+
+class ClintoArgumentParserException(Exception):
+    def __init__(self, parser, *args, **kwargs):
+        self.parser = parser
+
+
+def parse_args_monkeypatch(self, *args, **kwargs):
+    raise ClintoArgumentParserException(self)
+
+
+class BaseParser(object):
+
+    extensions = []
+    contains = ""
+
+    def __init__(self, script_path=None, script_source=None):
+        self.is_valid = False
+        self.error = ''
+        self.parser = None
+
+        self.script_path = script_path
+        self.script_source = script_source
+
+        if not self.check_valid():
+            return
+
+        self.extract_parser()
+
+        if self.parser is None:
+            return
+
+        self.process_parser()
+
+    def check_valid(self):
+        if self.script_path:
+            ext = os.path.splitext(os.path.basename(self.script_path))[1]
+            return ext in self.extensions and self.contains in self.script_source
```

Comment:
How are you importing from the common file in the standalone script? Is it installed as a module?

The problem is though we can test for a ArgumentParser we can only do that *after* we have the ArgumentParser. The check was to see how likely it would be to get a parser *before* getting it. 

Otherwise we'll have repeatedly process the script (for argparse, click, etc. etc.) until we hit a match. I was hoping to shortcut that... might not be possible.

How about rather than a blunt check it returns a heuristic 'score' for each parser, and then we can test them in order? That will cut down on processing for most things but allow it to work regardless.